File: src/TrackForge/Core/NameFormatter.cs
using System.Text;
using System.Text.RegularExpressions;

namespace TrackForge.Core;

/// <summary>Title Case rules and filename building, matching the F:\Music convention.</summary>
public static partial class NameFormatter
{
    // Deliberately small: the library capitalises "Of" and "On", so only true
    // connectors stay lowercase mid-title.
    private static readonly HashSet<string> LowerWords = new(StringComparer.OrdinalIgnoreCase)
        { "a", "an", "the", "and", "or", "nor", "but", "vs", "feat", "ft" };

    [GeneratedRegex(@"[<>:""/\\|?*\x00-\x1f]")]
    private static partial Regex IllegalChars();

    [GeneratedRegex(@"\s+")]
    private static partial Regex Whitespace();

    [GeneratedRegex(@"[^A-Za-z]")]
    private static partial Regex NonLetters();

    public static string TitleCase(string? input)
    {
        if (string.IsNullOrWhiteSpace(input)) return input ?? "";
        var words = Whitespace().Replace(input.Replace('_', ' ').Trim(), " ").Split(' ');
        var sb = new StringBuilder();

        for (int i = 0; i < words.Length; i++)
        {
            var w = words[i];
            if (w.Length == 0) continue;
            if (sb.Length > 0) sb.Append(' ');

            var bare = NonLetters().Replace(w, "");
            // Leave acronyms (B.Y.O.B., ADD) and deliberate inner caps (DDevil, iTunes) alone.
            bool acronym = bare.Length > 0 && bare.ToUpperInvariant() == bare;
            bool innerCaps = bare.Length > 1 && bare[1..] != bare[1..].ToLowerInvariant();
            if (acronym || innerCaps) { sb.Append(w); continue; }

            var lower = w.ToLowerInvariant();
            bool middle = i > 0 && i < words.Length - 1;
            if (middle && LowerWords.Contains(lower.Trim('.', ',', '(', ')')))
                sb.Append(lower);
            else
                sb.Append(char.ToUpperInvariant(lower[0])).Append(lower[1..]);
        }
        return sb.ToString();
    }

    public static string SafeFileName(string? s)
    {
        var cleaned = IllegalChars().Replace(s ?? "", "").Trim().TrimEnd('.');
        cleaned = Whitespace().Replace(cleaned, " ").Trim();
        return cleaned.Length == 0 ? "Untitled" : cleaned;
    }

    /// <summary>
    /// Build a filename from a pattern. Tokens: {track} {tracknum} {title}
    /// {artist} {albumartist} {album} {year}.
    /// </summary>
    public static string BuildFileName(Track t, string pattern, string extension)
    {
        var rawTrack = (t.TrackNumber ?? "").Split('/')[0].Trim();
        bool numeric = int.TryParse(rawTrack, out int trackNo) && trackNo > 0;

        var map = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["{track}"] = numeric ? trackNo.ToString("00") : "",
            ["{tracknum}"] = numeric ? trackNo.ToString() : "",
            ["{title}"] = TitleCase(t.Title),
            ["{artist}"] = TitleCase(t.Artist),
            ["{albumartist}"] = TitleCase(string.IsNullOrWhiteSpace(t.AlbumArtist) ? t.Artist : t.AlbumArtist),
            ["{album}"] = TitleCase(t.Album),
            ["{year}"] = t.Year ?? "",
        };

        var name = pattern;
        foreach (var (token, value) in map)
            name = Regex.Replace(name, Regex.Escape(token), value.Replace("$", "$$"),
                                 RegexOptions.IgnoreCase);

        name = Whitespace().Replace(name, " ").Trim(' ', '-', '_');
        if (name.Length == 0) name = TitleCase(t.Title);
        if (!extension.StartsWith('.')) extension = "." + extension;
        return SafeFileName(name) + extension;
    }

    /// <summary>Adds " (2)", " (3)"... until the path is free.</summary>
    public static string UniquePath(string desired)
    {
        if (!File.Exists(desired)) return desired;
        var dir = Path.GetDirectoryName(desired) ?? "";
        var stem = Path.GetFileNameWithoutExtension(desired);
        var ext = Path.GetExtension(desired);
        for (int n = 2; n < 1000; n++)
        {
            var candidate = Path.Combine(dir, $"{stem} ({n}){ext}");
            if (!File.Exists(candidate)) return candidate;
        }
        return Path.Combine(dir, $"{stem} ({Guid.NewGuid():N}){ext}");
    }
}

File: macOS/Sources/NameFormatter.swift
import Foundation

/// Title Case rules and filename building, matching the library convention.
enum NameFormatter {
    // Deliberately small: the library capitalises "Of" and "On", so only true
    // connectors stay lowercase mid-title.
    private static let lowerWords: Set<String> =
        ["a", "an", "the", "and", "or", "nor", "but", "vs", "feat", "ft"]

    /// `/` and `:` are what macOS actually forbids, but the Windows set is kept so
    /// a library shared between the two platforms names files identically.
    private static let illegal = CharacterSet(charactersIn: "<>:\"/\\|?*")
        .union(CharacterSet(charactersIn: UnicodeScalar(0)...UnicodeScalar(31)))

    static func titleCase(_ input: String?) -> String {
        guard let input, !input.isBlank else { return input ?? "" }

        let words = input.replacingOccurrences(of: "_", with: " ")
            .trimmed
            .split(separator: " ", omittingEmptySubsequences: true)
            .map(String.init)

        var out: [String] = []
        for (i, w) in words.enumerated() {
            let bare = w.filter { $0.isLetter }

            // Leave acronyms (B.Y.O.B., ADD) and deliberate inner caps
            // (DDevil, iTunes) alone.
            let acronym = !bare.isEmpty && bare == bare.uppercased()
            let innerCaps = bare.count > 1 && String(bare.dropFirst()) != bare.dropFirst().lowercased()
            if acronym || innerCaps { out.append(w); continue }

            let lower = w.lowercased()
            let middle = i > 0 && i < words.count - 1
            let stripped = lower.trimmingCharacters(in: CharacterSet(charactersIn: ".,()[]\"'"))

            if middle && lowerWords.contains(stripped) {
                out.append(lower)
            } else {
                // Capitalise the first LETTER, not the first character: a word
                // like "(deluxe" starts with a bracket, and upper-casing that
                // changes nothing, which is how "(deluxe Edition)" slipped through.
                if let idx = lower.firstIndex(where: { $0.isLetter }) {
                    out.append(lower[..<idx]
                        + lower[idx].uppercased()
                        + lower[lower.index(after: idx)...])
                } else {
                    out.append(lower)
                }
            }
        }
        return out.joined(separator: " ")
    }

    static func safeFileName(_ s: String?) -> String {
        var cleaned = (s ?? "")
            .components(separatedBy: illegal)
            .joined()
            .trimmed
        while cleaned.hasSuffix(".") { cleaned = String(cleaned.dropLast()) }
        cleaned = cleaned.split(separator: " ", omittingEmptySubsequences: true)
            .joined(separator: " ")
            .trimmed
        return cleaned.isEmpty ? "Untitled" : cleaned
    }

    /// Build a filename from a pattern. Tokens: {track} {tracknum} {title}
    /// {artist} {albumartist} {album} {year}.
    static func buildFileName(_ t: Track, pattern: String, extension ext: String) -> String {
        let rawTrack = (t.trackNumber.split(separator: "/").first.map(String.init) ?? "").trimmed
        let trackNo = Int(rawTrack) ?? 0
        let numeric = trackNo > 0

        let map: [(String, String)] = [
            ("{track}", numeric ? String(format: "%02d", trackNo) : ""),
            ("{tracknum}", numeric ? String(trackNo) : ""),
            ("{title}", titleCase(t.title)),
            ("{artist}", titleCase(t.artist)),
            ("{albumartist}", titleCase(t.albumArtist.isBlank ? t.artist : t.albumArtist)),
            ("{album}", titleCase(t.album)),
            ("{year}", t.year),
        ]

        var name = pattern
        for (token, value) in map {
            name = name.replacingOccurrences(
                of: token, with: value, options: .caseInsensitive)
        }

        name = name.split(separator: " ", omittingEmptySubsequences: true)
            .joined(separator: " ")
            .trimmingCharacters(in: CharacterSet(charactersIn: " -_"))
        if name.isEmpty { name = titleCase(t.title) }

        var extension_ = ext
        if !extension_.hasPrefix(".") { extension_ = "." + extension_ }
        return safeFileName(name) + extension_
    }

    /// Adds " (2)", " (3)"… until the path is free.
    static func uniquePath(_ desired: String) -> String {
        let fm = FileManager.default
        guard fm.fileExists(atPath: desired) else { return desired }

        let url = URL(fileURLWithPath: desired)
        let dir = url.deletingLastPathComponent()
        let stem = url.deletingPathExtension().lastPathComponent
        let ext = url.pathExtension

        for n in 2..<1000 {
            let candidate = dir
                .appendingPathComponent("\(stem) (\(n))")
                .appendingPathExtension(ext)
            if !fm.fileExists(atPath: candidate.path) { return candidate.path }
        }
        return dir
            .appendingPathComponent("\(stem) (\(UUID().uuidString))")
            .appendingPathExtension(ext).path
    }
}
